infrastructures/openshift/src/main/java/org/eclipse/che/workspace/infrastructure/openshift/provision/AsyncStorageProvisioner.java

```diff
@@ -0,0 +1,370 @@
+      throw new InfrastructureException(message);
+    }
+
+    if ("true".equals(osEnv.getAttributes().get(ASYNC_PERSIST_ATTRIBUTE))
+        && !isEphemeral(osEnv.getAttributes())) {
+      String message =
+          "Workspace configuration not valid: Asynchronous storage available only if attribute 'persistVolumes' set to false";
+      LOG.warn(message);
+      osEnv.addWarning(new WarningImpl(4200, message));
+      throw new InfrastructureException(message);
+    }
+
+    String namespace = identity.getInfrastructureNamespace();
+    KubernetesClient oc = clientFactory.create(identity.getWorkspaceId());
+
+    boolean isPvcExist =
+        oc.persistentVolumeClaims()
+            .inNamespace(namespace)
+            .list()
+            .getItems()
+            .stream()
+            .anyMatch(
+                (Predicate<PersistentVolumeClaim>)
+                    pvc -> pvc.getMetadata().getName().equals(configuredPVCName));
+
+    if (!isPvcExist) {
+      PersistentVolumeClaim pvc =
+          KubernetesObjectUtil.newPVC(
+              configuredPVCName, accessMode, pvcQuantity, pvcStorageClassName);
+      oc.persistentVolumeClaims().inNamespace(namespace).create(pvc);
+    }
+
+    String configMapName = namespace + ASYNC_STORAGE_CONFIG;
+    boolean isMapExist =
+        oc.configMaps()
+            .inNamespace(namespace)
+            .list()
+            .getItems()
+            .stream()
+            .anyMatch(
+                (Predicate<ConfigMap>)
+                    configMap -> configMap.getMetadata().getName().equals(configMapName));
+
+    if (!isMapExist) {
+      ConfigMap configMap = createConfigMap(namespace, configMapName, identity, osEnv);
+      oc.configMaps().inNamespace(namespace).create(configMap);
+    }
+
+    boolean isPodExist =
+        oc.pods()
+            .inNamespace(namespace)
+            .list()
+            .getItems()
+            .stream()
+            .anyMatch((Predicate<Pod>) pod -> pod.getMetadata().getName().equals(ASYNC_STORAGE));
+
+    if (!isPodExist) {
+      Pod pod = createStoragePod(namespace, configMapName);
+      oc.pods().inNamespace(namespace).create(pod);
+    }
+
+    boolean isServiceExist =
+        oc.services()
+            .inNamespace(namespace)
+            .list()
+            .getItems()
+            .stream()
+            .anyMatch(
+                (Predicate<Service>)
+                    service -> service.getMetadata().getName().equals(ASYNC_STORAGE));
+
+    if (!isServiceExist) {
+      Service service = createStorageService(namespace);
+      oc.services().inNamespace(namespace).create(service);
+    }
+  }
+
+  /** Get or create new pair of SSH keys, this is need for securing rsync connection */
+  private List<SshPairImpl> getOrCreateSshPairs(
+      RuntimeIdentity identity, OpenShiftEnvironment osEnv) throws InfrastructureException {
+    List<SshPairImpl> sshPairs;
+    try {
+      sshPairs = sshManager.getPairs(identity.getOwnerId(), "internal");
```

Comment:
Should we make  constant for "internal"?

Reply:
`internal` is very non-descriptive. Is this supposed to be pre-existing? Why not have a dedicated keypair for rsync?